### ISSUE TITLE
Handle override entries by columns in need instead of fixed column set

### ISSUE THREAD

**Issue by CaiMiao** [summertriangle-dev/sparklebox#7]
This is optional, but I need this to improve my translated name search suggestion.
It's crazy to maintain a whole table with things have been in the `name.csv`,
and overriding by column is a more light-weight way to set additional strings in case of need.

I'm planning to add another translate table instead of editting existing `override.csv`,
which could avoid handling merge conflict each time there's a new override entry,
and I also planning to add another custom optional column for new search string ( traditional chinese, this may totally same as `kanji` so in most case it's blank for most idols to save bytes ).

**Comment by CaiMiao** [comment]
network lags.

**Comment by CaiMiao** [comment]
Reopen to save resources.

**Comment by CaiMiao** [comment]
Sorry, changes to title is not correctly commited.

**Comment by summertriangle-dev** [comment]
I don't exactly understand what you want. An overlay on top of _data/private/overrides.csv?

Can you post some example files?

**Comment by CaiMiao** [comment]
maybe like this.

> chara_id,kanji,translated,translated_cht,conventional
> 1,千川ちひろ,千川ちひろ,
> 2,千川ちひろ,千川千寻,千川千尋
> 3,エトセトラ,エトセトラ,
> 18,ぴにゃこら太,ぴにゃこら太,
> 19,うさぎ,うさぎ,
> 20,千川ちひろ,千川ちひろ,
> 101,島村卯月,岛村卯月,
> 102,中野有香,中野有香,
> 103,水本ゆかり,水本紫,
> 104,福山舞,福山舞,
> 105,椎名法子,椎名法子,
> 106,今井加奈,今井加奈,
> 107,持田亜里沙,持田亚里沙,持田亞里沙
> 108,三村かな子,三村かな子,
> 109,奥山沙織,奥山沙織,
> 111,小日向美穂,小日向美穂,
> 112,緒方智絵里,緒方智絵里,
> 113,五十嵐響子,五十嵐響子,
> 115,櫻井桃華,櫻井桃華,
> 118,横山千佳,横山千佳,
> 119,関裕美,関裕美,
> 121,棟方愛海,棟方愛海,
> 122,藤本里奈,藤本里奈,
> 123,大原みちる,大原みちる,
> 126,一ノ瀬志希,一ノ瀬志希,
> 127,前川みく,前川みく,
> 129,松原早耶,松原早耶,
> 131,宮本フレデリカ,宫本芙蕾德莉卡,宮本芙蕾德莉卡
>

**Comment by CaiMiao** [comment]
Indexed by id, overlays name strings on demand. (without copying same structure as name.csv)
Maybe this is what I mean.

**Comment by CaiMiao** [comment]
override.csv would also benefits from it.

**Comment by CaiMiao** [comment]
Sorry I'm on mobile version.
Fromer example is not completed, I'm just showing a sample.
You should notice id 101,107,131. that's what I want to say.

**Comment by summertriangle-dev** [comment]
Ok. So I'll do this:

- Make `overrides.csv` more like an overlay. The schema can be different from `names.csv`, and we'll just replace individual keys like this:

> names.csv:  
```
"chara_id","kanji","kanji_spaced","kana_spaced","other","conventional"
1,"XX","YY","hello","ZZ","CC"
```

> overrides.csv:  
```
"chara_id","kanji","kanji_spaced","kana_spaced","conventional"
1,"AB","YY","ZZ","CC"
```

Which will then result in the name object:
`name_t(chara_id=1, kanji="AB", kanji_spaced="YY", other="hello", kana_spaced="ZZ", conventional="CC")`

- Allow `names.csv` to have incomplete rows. Those will just be filled out with the empty string. So you can have:

> names.csv:  
```
"chara_id","kanji","kanji_spaced","kana_spaced","conventional","optional"
2,"AB","XX","YY","ZZ"
3,"AB","XX","YY","ZZ","defined"
```

Which will then result in the name objects:
`name_t(chara_id=2, kanji="AB", kanji_spaced="XX",kana_spaced="YY", conventional="ZZ", optional="")`  
`name_t(chara_id=3, kanji="AB", kanji_spaced="XX",kana_spaced="YY", conventional="ZZ", optional="defined")`

**Comment by summertriangle-dev** [comment]
Please reopen if 4b09dfd40621ef835bda075e87f5a79cfb933a7a doesn't do it for you

**Comment by CaiMiao** [comment]
I'm just planning to add a `translated.csv` to going through this, and I'm copying codes to load csv same as override, and put `names.update(translated)` after `names.update(overrides)`, with following format:
```
chara_id,kanji,translated,translated_cht
1,千川ちひろ,千川ちひろ,
2,千川ちひろ,千川千寻,千川千尋
```
and returns error that table is not properly loaded as I defined to load `translated` key.

**Comment by CaiMiao** [comment]
At last I made it but the code looks untidy, because it's totally same as `for key in overrides` section,
anyway to unite them?

**Comment by CaiMiao** [comment]
Not fully done, I have this.
`TypeError: __new__() got an unexpected keyword argument 'translated'`

**Comment by CaiMiao** [comment]
Check my codes here: 3aad93168700e6077b6030f4408e70cdef6f6dd9

**Comment by summertriangle-dev** [comment]
names.csv has to contain all the column names used by overrides and translated csvs (you can leave them blank). This is so we can avoid re-creating all the `name_t`s without override content. 

Also the way you are applying your file is the best way (as far as I can see). You'll need two for loops anyway because overrides and translated are different data sets.

**Comment by CaiMiao** [comment]
Ty. If I'm going to control search result to avoid repeating entries (each updates would have individual result if multiple keywords set are matched), how should I do?
For now I'm adding asterick to notice this may be a repeating entry.

**Comment by summertriangle-dev** [comment]
You could possibly de-duplicate based on the character ID in JS.

**Comment by CaiMiao** [comment]
Thank you, I'm using a non-elegant method to kill repeated IDs. 
Check it if you get interested : 8954332bd58cec0da8906f628346680c1e6e312c

**Comment by CaiMiao** [comment]
Any way to specify a key column for individual csv to overlay with another key (eg. use `kanji` instead) ?
This could be a great help for overlaying those idols which are not actually implemented in game/truth.
Maintaining a csv each week is a pain :-(

**Comment by summertriangle-dev** [comment]
Good idea, implemented in f448e4f. 

If `chara_id` is specified as a negative number, then the row will get applied to the entry with a matching `kanji` value. (And because of how dicts work, each negative ID has to be unique. You can just pick a random number.)

**Comment by CaiMiao** [comment]
check searching on my site http://starlight.hyspace.io/ , ids are resulting undifined.
